File: src/plugin-kernel/ipc/mxpd.ts
import {
  FLAG_ACK,
  FLAG_CANCEL,
  FLAG_END,
  IPC_VERSION,
  MAX_PAYLOAD,
  MXPD_HEADER_BYTES,
  MXPD_MAGIC,
  RESERVED_FLAG_MASK,
  WINDOW_BYTES,
  WINDOW_FRAMES,
  ipcError,
  isKnownCodec,
  type IpcError,
} from "./constants";

export interface MxpdFrame {
  flags: number;
  streamId: number;
  seq: number;
  payload: Uint8Array;
}

export interface MxpdDecodeOk {
  frame: MxpdFrame;
  rest: Uint8Array;
}

function readU32be(bytes: Uint8Array, offset: number): number {
  return (
    ((bytes[offset] ?? 0) << 24) |
    ((bytes[offset + 1] ?? 0) << 16) |
    ((bytes[offset + 2] ?? 0) << 8) |
    (bytes[offset + 3] ?? 0)
  ) >>> 0;
}

function writeU32be(bytes: Uint8Array, offset: number, value: number): void {
  bytes[offset] = (value >>> 24) & 0xff;
  bytes[offset + 1] = (value >>> 16) & 0xff;
  bytes[offset + 2] = (value >>> 8) & 0xff;
  bytes[offset + 3] = value & 0xff;
}

function readU32le(bytes: Uint8Array, offset: number): number {
  return (
    (bytes[offset] ?? 0) |
    ((bytes[offset + 1] ?? 0) << 8) |
    ((bytes[offset + 2] ?? 0) << 16) |
    ((bytes[offset + 3] ?? 0) << 24)
  ) >>> 0;
}

function writeU32le(bytes: Uint8Array, offset: number, value: number): void {
  bytes[offset] = value & 0xff;
  bytes[offset + 1] = (value >>> 8) & 0xff;
  bytes[offset + 2] = (value >>> 16) & 0xff;
  bytes[offset + 3] = (value >>> 24) & 0xff;
}

export function encodeMxpd(frame: MxpdFrame): { ok: true; bytes: Uint8Array } | { ok: false; error: IpcError } {
  if ((frame.flags & RESERVED_FLAG_MASK) !== 0) {
    return { ok: false, error: ipcError("reserved-flag", `reserved MXPD flags set: ${frame.flags}`) };
  }
  if (frame.payload.length > MAX_PAYLOAD) {
    return { ok: false, error: ipcError("payload-too-large", `payload ${frame.payload.length} exceeds ${MAX_PAYLOAD}`) };
  }
  const bytes = new Uint8Array(MXPD_HEADER_BYTES + frame.payload.length);
  writeU32be(bytes, 0, MXPD_MAGIC);
  bytes[4] = IPC_VERSION;
  bytes[5] = frame.flags & 0xff;
  writeU32le(bytes, 6, frame.streamId >>> 0);
  writeU32le(bytes, 10, frame.seq >>> 0);
  writeU32le(bytes, 14, frame.payload.length);
  bytes.set(frame.payload, MXPD_HEADER_BYTES);
  return { ok: true, bytes };
}

export function decodeMxpd(bytes: Uint8Array): { ok: true; value: MxpdDecodeOk } | { ok: false; error: IpcError } {
  if (bytes.length < MXPD_HEADER_BYTES) {
    return { ok: false, error: ipcError("need-more", "incomplete MXPD header") };
  }
  const magic = readU32be(bytes, 0);
  if (magic !== MXPD_MAGIC) {
    return { ok: false, error: ipcError("bad-magic", `expected MXPD, got 0x${magic.toString(16)}`) };
  }
  const version = bytes[4] ?? 0;
  if (version !== IPC_VERSION) {
    return { ok: false, error: ipcError("unsupported-version", `version ${version} is not 1`) };
  }
  const flags = bytes[5] ?? 0;
  if ((flags & RESERVED_FLAG_MASK) !== 0) {
    return { ok: false, error: ipcError("reserved-flag", `reserved MXPD flags set: ${flags}`) };
  }
  const payloadLen = readU32le(bytes, 14);
  if (payloadLen > MAX_PAYLOAD) {
    return { ok: false, error: ipcError("payload-too-large", `payload_len ${payloadLen} exceeds ${MAX_PAYLOAD}`) };
  }
  if (bytes.length < MXPD_HEADER_BYTES + payloadLen) {
    return { ok: false, error: ipcError("truncated", "data frame payload is incomplete") };
  }
  return {
    ok: true,
    value: {
      frame: {
        flags,
        streamId: readU32le(bytes, 6),
        seq: readU32le(bytes, 10),
        payload: bytes.subarray(MXPD_HEADER_BYTES, MXPD_HEADER_BYTES + payloadLen),
      },
      rest: bytes.subarray(MXPD_HEADER_BYTES + payloadLen),
    },
  };
}

export function assertKnownCodec(codec: string): IpcError | null {
  if (!isKnownCodec(codec)) return ipcError("unknown-codec", `codec ${codec} is not a V1 codec`);
  return null;
}

export function canSend(unackedFrames: number, unackedBytes: number, nextPayloadBytes: number): IpcError | null {
  if (unackedFrames >= WINDOW_FRAMES || unackedBytes + nextPayloadBytes > WINDOW_BYTES) {
    return ipcError("window-exceeded", "unacked window is 32 frames or 8 MiB");
  }
  return null;
}

export const MXPD_FLAG = { END: FLAG_END, CANCEL: FLAG_CANCEL, ACK: FLAG_ACK } as const;

File: src-tauri/src/plugin_runtime/mxpd.rs
//! In-process MXPD data plane over pipes. No spawn, not in boot.

use std::collections::HashMap;
use std::io::{Read, Write};

use super::ipc::{
    assert_known_codec, can_send, decode_mxpd, encode_mxpd, IpcError, MxpdFrame, FLAG_ACK,
    FLAG_CANCEL, MXPD_HEADER_BYTES,
};

pub const DEFAULT_STREAMS_PER_GENERATION: usize = 8;

fn err(code: &'static str, message: impl Into<String>) -> IpcError {
    IpcError {
        code,
        message: message.into(),
    }
}

#[derive(Debug, Clone)]
struct StreamState {
    plugin_id: String,
    generation: u64,
    codec: String,
    unacked_frames: u32,
    unacked_bytes: u64,
    cancelled: bool,
}

#[derive(Default)]
pub struct DataPlane {
    streams: HashMap<u32, StreamState>,
}

impl DataPlane {
    pub fn open(
        &mut self,
        plugin_id: &str,
        generation: u64,
        stream_id: u32,
        codec: &str,
    ) -> Result<(), IpcError> {
        assert_known_codec(codec)?;
        if self.streams.contains_key(&stream_id) {
            return Err(err("stream-exists", format!("stream {stream_id} already open")));
        }
        let open_for_generation = self
            .streams
            .values()
            .filter(|stream| stream.plugin_id == plugin_id && stream.generation == generation)
            .count();
        if open_for_generation >= DEFAULT_STREAMS_PER_GENERATION {
            return Err(err(
                "stream-budget",
                format!("generation {generation} already has {DEFAULT_STREAMS_PER_GENERATION} streams"),
            ));
        }
        self.streams.insert(
            stream_id,
            StreamState {
                plugin_id: plugin_id.to_string(),
                generation,
                codec: codec.to_string(),
                unacked_frames: 0,
                unacked_bytes: 0,
                cancelled: false,
            },
        );
        Ok(())
    }

    pub fn revoke(&mut self, plugin_id: &str, generation: u64) {
        self.streams.retain(|_, stream| {
            !(stream.plugin_id == plugin_id && stream.generation == generation)
        });
    }

    pub fn write_frame(
        &mut self,
        writer: &mut impl Write,
        frame: &MxpdFrame,
    ) -> Result<(), IpcError> {
        let stream = self
            .streams
            .get_mut(&frame.stream_id)
            .ok_or_else(|| err("not-open", format!("stream {} is not open", frame.stream_id)))?;
        if stream.cancelled && frame.flags & FLAG_ACK == 0 {
            return Err(err("cancelled", "non-ACK frames after CANCEL are dropped"));
        }
        if frame.flags & FLAG_CANCEL != 0 {
            stream.cancelled = true;
        }
        if frame.flags & FLAG_ACK == 0 {
            can_send(stream.unacked_frames, stream.unacked_bytes, frame.payload.len() as u64)?;
            stream.unacked_frames += 1;
            stream.unacked_bytes += frame.payload.len() as u64;
        }
        write_mxpd_frame(writer, frame)
    }

    pub fn ack(&mut self, stream_id: u32, payload_bytes: u64) -> Result<(), IpcError> {
        let stream = self
            .streams
            .get_mut(&stream_id)
            .ok_or_else(|| err("not-open", format!("stream {stream_id} is not open")))?;
        if stream.unacked_frames == 0 {
            return Err(err("stale-ack", "no unacked frames"));
        }
        stream.unacked_frames -= 1;
        stream.unacked_bytes = stream.unacked_bytes.saturating_sub(payload_bytes);
        Ok(())
    }

    pub fn codec(&self, stream_id: u32) -> Option<&str> {
        self.streams.get(&stream_id).map(|stream| stream.codec.as_str())
    }
}

pub fn write_mxpd_frame(writer: &mut impl Write, frame: &MxpdFrame) -> Result<(), IpcError> {
    let bytes = encode_mxpd(frame)?;
    writer
        .write_all(&bytes)
        .map_err(|error| err("transport", error.to_string()))?;
    writer
        .flush()
        .map_err(|error| err("transport", error.to_string()))?;
    Ok(())
}

pub fn read_mxpd_frame(reader: &mut impl Read) -> Result<MxpdFrame, IpcError> {
    let mut header = [0_u8; MXPD_HEADER_BYTES];
    reader.read_exact(&mut header).map_err(|error| {
        if error.kind() == std::io::ErrorKind::UnexpectedEof {
            err("truncated", "incomplete MXPD header")
        } else {
            err("transport", error.to_string())
        }
    })?;
    let payload_len = u32::from_le_bytes([header[14], header[15], header[16], header[17]]) as usize;
    let mut frame = Vec::with_capacity(MXPD_HEADER_BYTES + payload_len);
    frame.extend_from_slice(&header);
    frame.resize(MXPD_HEADER_BYTES + payload_len, 0);
    reader
        .read_exact(&mut frame[MXPD_HEADER_BYTES..])
        .map_err(|error| {
            if error.kind() == std::io::ErrorKind::UnexpectedEof {
                err("truncated", "incomplete MXPD payload")
            } else {
                err("transport", error.to_string())
            }
        })?;
    let (decoded, _) = decode_mxpd(&frame)?;
    Ok(decoded)
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::io::Read;
    use std::thread;

    fn blob(seq: u32, payload: &[u8]) -> MxpdFrame {
        MxpdFrame {
            flags: 0,
            stream_id: 7,
            seq,
            payload: payload.to_vec(),
        }
    }

    #[test]
    fn send_without_open_is_rejected() {
        let mut plane = DataPlane::default();
        let (mut reader, mut writer) = std::io::pipe().expect("pipe");
        let error = plane
            .write_frame(&mut writer, &blob(1, b"nope"))
            .unwrap_err();
        assert_eq!(error.code, "not-open");
        drop(writer);
        let mut leftover = Vec::new();
        reader.read_to_end(&mut leftover).expect("drain");
        assert!(leftover.is_empty());
    }

    #[test]
    fn blob_frame_round_trips_on_a_pipe_after_open() {
        let mut plane = DataPlane::default();
        plane.open("com.mossx.notes", 1, 7, "blob-v1").expect("open");
        let (mut reader, mut writer) = std::io::pipe().expect("pipe");
        let peer = thread::spawn(move || read_mxpd_frame(&mut reader).expect("peer frame"));
        plane
            .write_frame(&mut writer, &blob(1, b"hello-blob"))
            .expect("write");
        drop(writer);
        let received = peer.join().expect("peer");
        assert_eq!(received.stream_id, 7);
        assert_eq!(received.payload, b"hello-blob");
        plane.ack(7, 10).expect("ack");
        assert_eq!(plane.codec(7), Some("blob-v1"));
    }

    #[test]
    fn cancel_drops_later_data_frames() {
        let mut plane = DataPlane::default();
        plane.open("com.mossx.notes", 1, 7, "blob-v1").expect("open");
        let (mut reader, mut writer) = std::io::pipe().expect("pipe");
        plane
            .write_frame(
                &mut writer,
                &MxpdFrame {
                    flags: FLAG_CANCEL,
                    stream_id: 7,
                    seq: 1,
                    payload: Vec::new(),
                },
            )
            .expect("cancel");
        let error = plane
            .write_frame(&mut writer, &blob(2, b"after-cancel"))
            .unwrap_err();
        assert_eq!(error.code, "cancelled");
        drop(writer);
        let first = read_mxpd_frame(&mut reader).expect("cancel frame");
        assert_eq!(first.flags, FLAG_CANCEL);
        let mut leftover = Vec::new();
        reader.read_to_end(&mut leftover).expect("drain");
        assert!(leftover.is_empty());
    }

    #[test]
    fn window_blocks_unacked_frames() {
        let mut plane = DataPlane::default();
        plane.open("com.mossx.notes", 1, 7, "log-v1").expect("open");
        let (mut reader, mut writer) = std::io::pipe().expect("pipe");
        for seq in 1..=32 {
            plane
                .write_frame(&mut writer, &blob(seq, b"x"))
                .expect("within window");
        }
        let error = plane.write_frame(&mut writer, &blob(33, b"x")).unwrap_err();
        assert_eq!(error.code, "window-exceeded");
        drop(writer);
        let mut leftover = Vec::new();
        reader.read_to_end(&mut leftover).expect("drain");
        assert!(!leftover.is_empty());
    }

    #[test]
    fn revoked_generation_cannot_write() {
        let mut plane = DataPlane::default();
        plane.open("com.mossx.notes", 1, 7, "blob-v1").expect("open");
        let (mut reader, mut writer) = std::io::pipe().expect("pipe");
        plane.revoke("com.mossx.notes", 1);
        let error = plane
            .write_frame(&mut writer, &blob(1, b"stale"))
            .unwrap_err();
        assert_eq!(error.code, "not-open");
        assert!(plane.codec(7).is_none());
        drop(writer);
        let mut leftover = Vec::new();
        reader.read_to_end(&mut leftover).expect("drain");
        assert!(leftover.is_empty());
    }
}
